controllers/mirroring/mirroring_controller.go

```diff
@@ -0,0 +1,478 @@
+			enqueueConfigMapRequest,
+			generationChangePredicate,
+		).
+		Complete(r)
+}
+
+//+kubebuilder:rbac:groups=ocs.openshift.io,resources=storageclusterpeers;storageconsumers,verbs=get;list;watch
+//+kubebuilder:rbac:groups=core,resources=configmaps,verbs=get;list;watch
+//+kubebuilder:rbac:groups=core,resources=configmaps/finalizers,verbs=update
+//+kubebuilder:rbac:groups=core,resources=secrets,verbs=get;list;watch;create;update;delete
+//+kubebuilder:rbac:groups=ceph.rook.io,resources=cephrbdmirrors,verbs=get;list;watch;create;update;delete
+//+kubebuilder:rbac:groups=ceph.rook.io,resources=cephblockpools;cephblockpoolradosnamespaces,verbs=get;list;watch;create;update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *MirroringReconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	r.ctx = ctx
+	r.log = log.FromContext(ctx, "StorageClientMapping", request.NamespacedName)
+	r.log.Info("Starting reconcile")
+
+	clientMappingConfig := &corev1.ConfigMap{}
+	clientMappingConfig.Name = request.Name
+	clientMappingConfig.Namespace = request.Namespace
+
+	if err := r.get(clientMappingConfig); err != nil {
+		if k8serrors.IsNotFound(err) {
+			r.log.Info("ConfigMap %s not found. Ignoring since object must be deleted.")
+			return ctrl.Result{}, nil
+		}
+		r.log.Error(err, "Failed to get ConfigMap.")
+		return ctrl.Result{}, err
+	}
+	return r.reconcilePhases(clientMappingConfig)
+}
+
+func (r *MirroringReconciler) reconcilePhases(clientMappingConfig *corev1.ConfigMap) (ctrl.Result, error) {
+
+	disableMirroring := false
+	disableMirroring = disableMirroring || !clientMappingConfig.DeletionTimestamp.IsZero()
+	disableMirroring = disableMirroring || clientMappingConfig.Data == nil || len(clientMappingConfig.Data) < 1
+
+	if !disableMirroring {
+		if controllerutil.AddFinalizer(clientMappingConfig, mirroringFinalizer) {
+			r.log.Info("Finalizer not found for ConfigMap. Adding finalizer.")
+			if err := r.update(clientMappingConfig); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to update ConfigMap: %v", err)
+			}
+		}
+	}
+
+	// Fetch the StorageClusterPeer instance
+	storageClusterPeerList := &ocsv1.StorageClusterPeerList{}
+	if err := r.list(storageClusterPeerList, client.InNamespace(clientMappingConfig.Namespace)); err != nil {
+		r.log.Error(err, "Failed to get StorageClusterPeer.")
+		return ctrl.Result{}, err
+	}
+	if len(storageClusterPeerList.Items) != 1 {
+		return ctrl.Result{}, fmt.Errorf("expected 1 StorageClusterPeer but got %d", len(storageClusterPeerList.Items))
+	}
+
+	storageClusterPeer := &storageClusterPeerList.Items[0]
+
+	if storageClusterPeer.Status.State != ocsv1.StorageClusterPeerStatePeered {
+		r.log.Info(
+			"waiting for StorageClusterPeer to be in Peered state",
+			"StorageClusterPeer",
+			storageClusterPeer.Name,
+		)
+		return ctrl.Result{Requeue: true}, nil
```

Comment:
This will create an immediate requeue, I would suggest to add a requeue backoff (maybe 1-5 seconds) to give some time for other process to complete